### ISSUE TITLE
Tone Equalizer does not utilize GPU

### ISSUE THREAD

**Issue by alexvanderberkel** [darktable-org/darktable#4805]
**Describe the bug**
I just noticed that the Tone Equalizer module does not utilize the capability of the GPU if OpenCL is activated. Instead of using the GPU it is using the CPU and, thus, slowing down the pixelpipe. I am not sure if this by intention or a limitation of this module. Below you can see which module took the longest for rendering. 

**To Reproduce**
run darktable with darktable -d perf
edit any image and note the output in the terminal

**Expected behavior**

**Screenshots**
163,913255 [dev_process_preview] pixel pipeline processing took 0,112 secs (0,135 CPU)
164,266260 [dev_pixelpipe] took 0,351 secs (1,786 CPU) initing base buffer [full]
164,366978 [dev_pixelpipe] took 0,101 secs (0,185 CPU) processed `lens correction' on GPU, blended on GPU [full]
164,369589 [dev_pixelpipe] took 0,003 secs (0,000 CPU) processed `crop and rotate' on GPU, blended on GPU [full]
164,628985 [dev_pixelpipe] took 0,259 secs (1,248 CPU) processed `tone equalizer' on CPU, blended on CPU [full]
164,642569 [dev_pixelpipe] took 0,014 secs (0,009 CPU) processed `input color profile' on GPU, blended on GPU [full]
164,783572 [dev_pixelpipe] took 0,141 secs (0,126 CPU) processed `contrast equalizer' on GPU, blended on GPU [full]
164,800374 [dev_pixelpipe] took 0,017 secs (0,008 CPU) processed `sharpen' on GPU, blended on GPU [full]
164,937148 [dev_pixelpipe] took 0,137 secs (0,116 CPU) processed `local contrast' on GPU, blended on GPU [full]
164,940186 [dev_pixelpipe] took 0,003 secs (0,000 CPU) processed `output color profile' on GPU, blended on GPU [full]
164,958459 [dev_pixelpipe] took 0,018 secs (0,054 CPU) processed `display encoding' on CPU, blended on CPU [full]
164,960713 [dev_process_image] pixel pipeline processing took 1,159 secs (3,669 CPU)

**Platform (please complete the following information):**
Ubuntu 20.04
darktable 3.1.0-git1300.75abcb81c

**Comment by jenshannoschwalm** [comment]
Just have a look at the modules's source, you will not find a cl path (yet) :-)

**Comment by alexvanderberkel** [comment]
True! Thanks for the hint. It is cleary marked as a todo. Hence, it is not a bug rather a feature request :)

**Comment by aurelienpierre** [comment]
Thinking about an OpenCL version of tone equalizer gives me gastric acidity. Interleaving the GUI and processing threads on CPU was already pushing darktable core in a way it has not been designed for, it took me 3 months of work. But convolving GPU in the mix…

I yearn for the GPU performance in tone equalizer, don't get me wrong, that thing is currently slow as hell. But, that job will drive me crazy.

**Comment by alexvanderberkel** [comment]
No worries! I love the module already a lot and we don't want you to go crazy! If it would be easy, you would already have implemented it ;)

**Comment by da-phil** [comment]
I'm going to pick up the work and provide a OpenCL code path for ToneEqualizer, given that ColorEqualizer is similar in some regards and already has a OpenCL code path, it's maybe not so much work.

**Comment by da-phil** [comment]
@TurboGit can you pls re-open the issue.

**Comment by TurboGit** [comment]
Done.

**Comment by da-phil** [comment]
Thanks @TurboGit 

First step of getting a full OpenCL implementation for the tone equalizer is done in https://github.com/darktable-org/darktable/pull/22042, by making the bilinear resampling kernels reusable, as they are also going to be needed by the tone equalizer.

**Comment by da-phil** [comment]
And here is a first draft for an opencl implementation: https://github.com/darktable-org/darktable/pull/22051

It's still a draft because it needs more testing and I want to fully address the implications mentioned by AP see https://github.com/darktable-org/darktable/issues/4805#issuecomment-619535525
